pkg/reconciler/pipelinerun/resources/pipelinerunstate.go

```diff
@@ -0,0 +1,319 @@
+	for _, rprt := range state {
+		allTasks = append(allTasks, rprt.PipelineTask.Name)
+		switch {
+		case rprt.IsSuccessful():
+			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
+		case rprt.Skip(state, dag):
+			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
+			skipTasks = append(skipTasks, v1beta1.SkippedTask{Name: rprt.PipelineTask.Name})
+			// At least one is skipped and no failure yet, mark as completed
+			if reason == v1beta1.PipelineRunReasonSuccessful.String() {
+				reason = v1beta1.PipelineRunReasonCompleted.String()
+			}
+		case rprt.IsCancelled():
+			cancelledTasks++
+			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
+			if reason != v1beta1.PipelineRunReasonFailed.String() {
+				reason = v1beta1.PipelineRunReasonCancelled.String()
+			}
+		case rprt.IsFailure():
+			withStatusTasks = append(withStatusTasks, rprt.PipelineTask.Name)
+			failedTasks++
+			reason = v1beta1.PipelineRunReasonFailed.String()
+		}
+	}
+
+	if reflect.DeepEqual(allTasks, withStatusTasks) {
+		status := corev1.ConditionTrue
+		if failedTasks > 0 || cancelledTasks > 0 {
+			status = corev1.ConditionFalse
+		}
+		logger.Infof("All TaskRuns have finished for PipelineRun %s so it has finished", pr.Name)
+		return &apis.Condition{
+			Type:   apis.ConditionSucceeded,
+			Status: status,
+			Reason: reason,
+			Message: fmt.Sprintf("Tasks Completed: %d (Failed: %d, Cancelled %d), Skipped: %d",
+				len(allTasks)-len(skipTasks), failedTasks, cancelledTasks, len(skipTasks)),
+		}
+	}
+
+	// Hasn't timed out; not all tasks have finished.... Must keep running then....
+	// transition pipeline into stopping state when one of the tasks(dag/final) cancelled or one of the dag tasks failed
+	// for a pipeline with final tasks, single dag task failure does not transition to interim stopping state
+	// pipeline stays in running state until all final tasks are done before transitioning to failed state
+	if cancelledTasks > 0 || (failedTasks > 0 && state.checkTasksDone(dfinally)) {
+		reason = v1beta1.PipelineRunReasonStopping.String()
+	} else {
+		reason = v1beta1.PipelineRunReasonRunning.String()
+	}
+	return &apis.Condition{
+		Type:   apis.ConditionSucceeded,
+		Status: corev1.ConditionUnknown,
+		Reason: reason,
+		Message: fmt.Sprintf("Tasks Completed: %d (Failed: %d, Cancelled %d), Incomplete: %d, Skipped: %d",
+			len(withStatusTasks)-len(skipTasks), failedTasks, cancelledTasks, len(allTasks)-len(withStatusTasks), len(skipTasks)),
+	}
+}
+
+func (state PipelineRunState) GetSkippedTasks(pr *v1beta1.PipelineRun, d *dag.Graph) []v1beta1.SkippedTask {
+	skipped := []v1beta1.SkippedTask{}
+	for _, rprt := range state {
+		if rprt.Skip(state, d) {
+			skipped = append(skipped, v1beta1.SkippedTask{Name: rprt.PipelineTask.Name})
+		}
+	}
+	return skipped
+}
+
+func (state PipelineRunState) GetTaskRunsStatus(pr *v1beta1.PipelineRun) map[string]*v1beta1.PipelineRunTaskRunStatus {
```

Comment:
nice - i like that none of these have pointer receivers so we don't have to worry about side effects :D